internal/config/variables/variables.go

```diff
@@ -329,23 +368,168 @@ func LoadEnvValues(vars map[string]*Variable) ([]*pb.Variable, hcl.Diagnostics)
+	}
+
+	// Build our watcher
+	ch := make(chan *appconfig.UpdatedConfig)
+	w, err := appconfig.NewWatcher(append(
+		append([]appconfig.Option{}, dynamicOpts...),
+		appconfig.WithLogger(log),
+		appconfig.WithNotify(ch),
+	)...)
+	if err != nil {
+		diags = append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Error initializing configuration watcher",
+			Detail:   err.Error(),
+			Subject: &hcl.Range{
+				Filename: "waypoint.hcl",
+			},
+		})
+		return nil, diags
+	}
+	defer w.Close()
+
+	// We have some variables we need to fetch dynamic values for.
+	configVars := make([]*pb.ConfigVar, 0, len(dynamicVars))
+	for k, v := range dynamicVars {
+		configVars = append(configVars, &pb.ConfigVar{
+			Name: k,
+			Value: &pb.ConfigVar_Dynamic{
+				Dynamic: v.Default.Value.EncapsulatedValue().(*pb.ConfigVar_DynamicVal),
+			},
+
+			// This is set to true on purpose because it forces the appconfig
+			// watcher to give us an easier to consume format (struct vs
+			// array of key=value strings for env vars).
+			NameIsPath: true,
+		})
+	}
+
+	// Send our variables. Purposely ignore the error return value because
+	// it can only ever be a context cancellation which we pick up in the
+	// select later.
+	w.UpdateVars(ctx, configVars)
+
+	// Wait for values.
+	select {
+	case <-ctx.Done():
+		diags = append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Cancellation while waiting for configuration.",
+			Detail:   ctx.Err().Error(),
+			Subject: &hcl.Range{
+				Filename: "waypoint.hcl",
+			},
+		})
+		return nil, diags
+
+	case config := <-ch:
+		var result []*pb.Variable
+		for _, f := range config.Files {
+			result = append(result, &pb.Variable{
+				Name: f.Path,
+				Value: &pb.Variable_Str{
+					Str: string(f.Data),
+				},
+				Source: &pb.Variable_Dynamic{},
+			})
+		}
+
+		return result, diags
+	}
+}
+
 // EvaluateVariables evaluates the provided variable values and validates their
 // types per the type declared in the waypoint.hcl for that variable name.
 // The order in which values are evaluated corresponds to their precedence, with
 // higher precedence values overwriting lower precedence values.
+//
 // The supplied map of *Variable should be all defined variables (currently
 // comes from decoding all variable blocks within the waypoint.hcl), and
 // is used to validate types and that all variables have at least one
 // assigned value.
 func EvaluateVariables(
+	log hclog.Logger,
 	pbvars []*pb.Variable,
 	vs map[string]*Variable,
-	log hclog.Logger,
 ) (Values, hcl.Diagnostics) {
 	var diags hcl.Diagnostics
 	iv := Values{}
 
 	for v, def := range vs {
+		// Do not allow dynamic values as default values since they aren't valid.
+		// Dynamic values should be evaluated and overridden by LoadDynamicDefaults
+		// and provided via pbvars. If not, then an unset error will be created.
+		if def.Default != nil && def.Default.Value.Type() == dynamic.Type {
+			continue
+		}
+
```

Comment:
Yeah not a big deal, just makes UX nicer. Since the dynamic type is an encapsulated value it can't be used for anything so its basically a poison pill. If anyone tried to use it for any values it'd just give an obscure HCL type error message. This just helps prevent that.